dali/python/nvidia/dali/eager.py

```diff
@@ -0,0 +1,242 @@
+        device_param = input_device if input_device else 'cpu'
+
+    sep_pos = device_param.find(':')
+
+    # Separate device and device_id.
+    if sep_pos != -1:
+        device = device_param[:sep_pos]
+        device_id = int(device_param[sep_pos + 1:])
+    else:
+        device = device_param
+        device_id = 0
+
+    if device == 'cpu' and input_device == 'gpu':
+        raise ValueError("An operator with device='cpu' cannot accept GPU inputs.")
+
+    if device != 'cpu' and device != 'gpu':
+        raise ValueError(f"Incorrect device type '{device}'.")
+
+    if input_device == 'cpu' and device == 'gpu':
+        if op_name in _ops._mixed_ops:
+            device = 'mixed'
+        else:
+            raise ValueError(f"Operator '{wrapper_name}' not registered for mixed.")
+
+    return device, device_id
+
+
+def _disqualify_arguments(op_name, kwargs, disqualified_args):
+    for key in disqualified_args:
+        if key in kwargs:
+            raise RuntimeError(f"Argument '{key}' is not supported by eager operator '{op_name}'.")
+
+
+def _choose_batch_size(inputs, batch_size):
+    """Returns batch size based on inputs and batch_size parameter."""
+
+    if len(inputs) > 0:
+        input_batch_size = len(inputs[0])
+
+        if batch_size == -1:
+            batch_size = input_batch_size
+
+        if input_batch_size != batch_size:
+            raise ValueError(
+                f"Requested batch_size={batch_size}, but input 0 has batch_size={input_batch_size}")
+
+    if batch_size == -1:
+        raise RuntimeError(
+            "Operators with no inputs need to have 'batch_size' parameter specified.")
+
+    return batch_size
+
+
+def _prep_inputs(inputs, batch_size):
+    inputs = list(inputs)
+
+    for i, input in enumerate(inputs):
+        if not isinstance(input, (_tensors.TensorListCPU, _tensors.TensorListGPU)):
+            inputs[i] = _transform_data_to_tensorlist(input, batch_size)
+
+    return inputs
+
+
+def _prep_kwargs(kwargs, batch_size):
+    for key, value in kwargs.items():
+        kwargs[key] = _Classification(value, f'Argument {key}', arg_constant_len=batch_size).data
+
+    return kwargs
+
+
+def _wrap_stateless(op_class, op_name, wrapper_name):
+    def wrapper(*inputs, **kwargs):
+        _disqualify_arguments(wrapper_name, kwargs, _wrap_stateless.disqualified_arguments)
+
+        # Preprocess kwargs to get batch_size.
+        batch_size = _choose_batch_size(inputs, kwargs.pop('batch_size', -1))
+        kwargs = _prep_kwargs(kwargs, batch_size)
+        init_args, call_args = _ops._separate_kwargs(kwargs, _tensors.TensorListCPU)
+
+        # Preprocess inputs, try to convert each input to TensorList.
+        inputs = _prep_inputs(inputs, batch_size)
+
+        init_args['max_batch_size'] = batch_size
+        init_args['device'], init_args['device_id'] = _choose_device(
+            op_name, wrapper_name, inputs, kwargs.get('device'))
+
+        key = op_name + str(sorted(init_args.items()))
```

Comment:
Hmm, this might be a bit problematic. I think we need to extend it to cover at least the types, dimensionality and possibly layouts of the call args (inputs and argument intpus). The operators do not expect those to change and using only init args as key may lead to us reusing the operator in such way. 
Do we have checks for that in debug mode? I think we do.

Reply:
Done, extended the key by dtype, layout and shape for each call arg.